src/usched.cpp

```diff
@@ -0,0 +1,236 @@
+/*
+User space qemu process cpu scheduler based on cpu_usage
+*/
+#include <algorithm>
+#include <chrono>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <mutex>
+#include <queue>
+#include <sched.h>
+#include <sstream>
+#include <string>
+#include<linux/sched.h>
+
+#include "usched.hpp"
+#include "cpu.h"
+#include "hfi.h"
+
+#define MAX_HISTORY 10
+#define test 0
+
+namespace fs = std::filesystem;
+
+const static int nprocs = get_nprocs_conf();
+const static int CLK = sysconf(_SC_CLK_TCK);
+
+static std::unordered_map<pid_t, struct pid_info *>
+    pool; //<<pid>,pid_info/tid_info>
+static std::mutex pool_mutex;
+static unsigned long long *cpu_prev_stat;
+static struct core_info *tmp_core_map;
+static unsigned int THREASHOLD = 100;
+
+// insert or update tid
+
+bool upsert_to_monitor_pool(pid_t qemu_id, pid_t tid,
+                            struct pid_info *_pid_info)
+{
+    if (qemu_id == tid)
+        return false;
+    std::lock_guard<std::mutex> lock(pool_mutex);
+    if (qemu_id == tid) {
+        return false;
+    }
+
+    std::string path = "/proc/" + std::to_string(qemu_id) + "/task/" +
+                       std::to_string((tid)) + "/stat";
+    std::ifstream fs(path);
+    std::stringstream buffer;
+    buffer << fs.rdbuf();
+    std::string str;
+    std::vector<std::string> data;
+    int idx = 0, stateidx = -1;
+    std::string state[] = {"R", "S", "D", "Z", "T", "t",
+                           "W", "X", "x", "K", "W", "P"};
+    while (buffer >> str) {
+        data.emplace_back(str);
+        if (stateidx == -1) {
+            for (std::string s : state)
+                if (s.compare(str) == 0)
+                    stateidx = idx;
+        }
+        idx++;
+    }
+    buffer.clear();
+    fs.clear();
+    fs.close();
+
+    if (stateidx == -1) {
+        return false;
+    }
+    // calc process usage
+    int last_cpu, utime, stime;
+    last_cpu = stoi(data[stateidx + TID_STAT_ITEM::PROCESSOR]);
+    utime = stoull(data[stateidx + TID_STAT_ITEM::UTIME]);
+    stime = stoull(data[stateidx + TID_STAT_ITEM::STIME]);
+
+    if (_pid_info == NULL) { // insert
+        struct pid_info *tmp = new struct pid_info(
```

Comment:
it'd be better to use `struct pid_info` instead of `struct pid_info *` in the pool so you don't need to new here. And new again for `pid_usage_info` is bad practice.